# Geocoding Organisations

For a number of planned analyses, we require the latitude and longitude of the organisations in the data, not just their names and addresses. To do this we will first use a geocoding service to identify lat lng from address strings.

In [226]:
import pickle
import ratelim
import json
import numpy as np
import os
import pandas as pd
import psycopg2
from sqlalchemy import create_engine

from geopy.geocoders import GoogleV3
from ipywidgets import FloatProgress
from IPython.display import display

key = os.environ.get('GOOGLE_MAPS_KEY')

In [103]:
# Google geocoder
geocoder = GoogleV3(api_key=key)

In [3]:
# Read in config file with DB params
with open('../scripts/config.json') as f:
    conf = json.load(f)
    
# Define a connection string
conn_string = 'host={} dbname={} user={} password={}'.format(conf.get('host'),
                                                             conf.get('database'),
                                                             conf.get('user'),
                                                             conf.get('passw'))

# Create a connection object
conn = psycopg2.connect(conn_string)

# SQL string that unpacks nested JSON arrays
sql_str = """
SELECT id, name, addresses from gtr.organisations;
"""

# chunksize returns an iterator that reads chunk number of rows at a time
df = pd.read_sql(sql_str,
                 conn)

In [5]:
def country(x):
    try:
        return x.get('address', None)[0].get('country', None).lower()
    except:
        return None

In [152]:
# API rate limits with billing enabled
@ratelim.greedy(5, 1)
def geocode(address_dict):
    # Get the list out of the dict
    address_list = address_dict['address']

    # If no country need to assume UK
    if address_list and address_list[0].get('country', 'uk').lower() in uk_names:
        try:
            pc = address_dict['address'][0]['postCode']
            return geocoder.geocode(pc)
        except:
            address = address_str(address_dict)
            if address:
                return geocoder.geocode(address)
            else:
                return None
    else:
        address = address_str(address_dict, uk=False)
        if address:
            g = geocoder.geocode(address)
        else:
            return None

In [105]:
def get_add_part(address_entry, part):
    try:
        return address_entry['address'][0].get('{}'.format(part), None)
    except:
        return None

In [106]:
def postcode(address_entry):
    return get_add_part(address_entry, 'postcode')

In [107]:
def country(address_entry):
    c = get_add_part(address_entry, 'country')
    if c == None:
        return 'uk'
    else:
        return c.lower()

In [252]:
def line1(address_entry):
    return get_add_part(address_entry, 'line1')

In [264]:
def address_str(address_dict, uk=True):
    
    uk_address_keys = [
    'line1',
    'line2',
    'line3',
    'county',
    'postCode',
    ]
    
    other_address_keys = [
        'line1',
        'line2',
        'line3',
        'line4',
        'line5'
        'county',
        'country',
    ]
    
    if uk:
        address_keys = uk_address_keys
    else:
        address_keys = other_address_keys
    
    l = []
    address_list = address_dict.get('address', None)
    if address_list is None:
        return None
    try:
        if address_list[0] is not None:
            [l.append(address_list[0][key]) if key in address_keys else '' for key in address_list[0]]            
    except:
        return None
    
    return ", ".join(l)

In [109]:
def check_address(address_dict):
    address_list = address_dict.get('address', None)
    if address_list is None:
        return False
    try:
        if address_list[0] is not None:
            return True
    except:
        return False

In [390]:
# Function to get tuple by position, returning
# None where there isn't a valid entry
def get_tuple_value(x, pos):
    try:
        return x[pos]
    except TypeError:
        return None

In [79]:
# All the ways people put down UK in the country field
uk_names = ['united kingdom',
 'uk',
 'wales',
 'england',
 'scotland',
 'united kingom',
 'northern ireland',
 'united kindgom',
 'buckinghamshire',
 'yorkshire',
 'london',
 'u.k.',
 'united ,kingdom',
 'berkshire',
 'jersey',
 ' united kingdom',
 'u.k',
 '|united kingdom']

In [11]:
df['postcode'] = df.addresses.map(postcode)
n_postcodes = df.postcode.count()
n_no_postcodes = len(df.postcode) - n_postcodes
print('{} organisations with postcodes, {} without'.format(n_postcodes, n_no_postcodes))

18266 organisations with postcodes, 7299 without


We will be unable to geocode 7,299 organisations by using only a postcode in the geocode request. Lets check how many of those without postcodes, also lack other address details. For instance, organisations from outside the UK would lack a postcode, but can still be geocoded using a string lookup.

In [12]:
# Full dataframe check
df.addresses.map(check_address).value_counts()

True     21205
False     4360
Name: addresses, dtype: int64

In [13]:
# Subset of those without postcodes
df[df.postcode.isnull()].addresses.map(check_address).value_counts()

False    4360
True     2939
Name: addresses, dtype: int64

Of the 7,299 organisations without a postcode, 4,360 also have no address data. This leaves 2,939 organisations with no postcode but some data in the address field. Therefore the maximum number of geocoded organisations will be 21,205. When geocoding, we will default to the use of a postcode where possible, and fallback to an address string where a postcode isn't present.

In [387]:
"""latlng = []
num_addresses = df.addresses.shape[0]
f = FloatProgress(min=0, max=num_addresses)
display(f)

for address in df.addresses:
    g = geocode(address)
    try:
        latlng.append((g.latitude, g.longitude))
    except AttributeError:
        latlng.append(None)
    f.value += 1
"""   

In [394]:
# Save as a pickled object
pickle.dump( latlng, open( "latlng.p", "wb" ) )

In [34]:
df_latlng = pd.DataFrame(np.asarray(latlng))

In [480]:
df_result = df.merge(df_latlng, left_index=True, right_index=True)
df_result.rename(columns={0: 'latlng'}, inplace=True)
df_result = df_result[df_result.latlng.notnull()]
df_result[['lat', 'lng']] = df_result.latlng.apply(pd.Series)
df_result.drop(['postcode', 'addresses', 'latlng'], axis=1, inplace=True)

In [481]:
df_result

id  \
0      BAB46AB6-90FC-4145-B885-0153BEA15A73   
2      BCD9069A-6739-4039-86DB-042CB13254CB   
3      B69FACF5-5AB9-4D97-88CD-0A38C32935FF   
4      B76FE143-1E07-40B0-8932-028207296A78   
5      B80095E4-9D0F-4398-8A0E-07EB418EFFF6   
6      B966BD7A-F468-4C74-A694-0842CB8F884F   
8      BA32CA96-1A26-415F-B3B2-0DC1DDFCE74F   
9      BA59F886-138D-41EB-B9B1-0C79E45C61D3   
10     BA7D8BB2-4B17-4D57-AC78-001BA0F40F77   
11     C059DB15-1836-430B-A993-0759206ABBB8   
13     C066B655-E1BA-447A-A432-01FB6725E320   
14     C2B3EE68-0D08-4321-964A-02108F67A3F8   
15     C30DF86F-9DD8-4A78-97BA-067527E5E391   
18     C42E852D-5D31-4DE8-8E91-04CF66DA9DE2   
19     C59C7E94-5DD5-4DCA-8586-09189A61F279   
20     C8A76B9E-4F39-490B-BCCF-009F592A756C   
21     C9069C42-693D-4C9C-B66B-0D968C7277AA   
22     C9424089-9FC7-4BD3-8C9A-0E38B885E2CE   
23     D04848C3-4132-45F1-8841-06DB92EB1E87   
24     D068E964-7B96-42CA-9F67-00BE1677AE7F   
25     D080060B-4CE4-495E-8C47-07BFD74CB8E0   
26     D3BC1D1D-CEC8-4A05-A0B0-0D6167F5A0B0   
27     D3E9EF03-CF8E-4251-BCF4-01E09BFADDA8   
28     D42BD99F-AD31-43D6-8FBE-0B96F1E266D3   
29     CA10F5D0-8E26-4790-BF21-025009DDA322   
32     D5542D2A-CB08-48B1-91A6-03A59EA2B30F   
33     D5F4323F-A8DF-4869-B0F0-0EB45A7A7F95   
34     D6C3234E-5048-4AA0-A116-0CAC51569B2D   
36     D72C440C-6E4E-4FEE-94D9-0652BB970C3B   
38     DFFCC0DB-5473-43BE-AC92-06CA6F272331   
...                                     ...   
25530  621FCA67-9259-4CBE-AECE-D7D4C4133C54   
25531  82DA3271-1C71-474F-BF1B-D122CF76DAFF   
25532  839FAFB1-DF5C-46B3-96AD-CF18D4BC7AFC   
25533  84E231EE-EC82-44AC-9838-C30021DDA51A   
25534  84F84A6B-60BD-4285-B4D0-E334D8B8938A   
25535  99A606D5-9E6E-4968-B976-C1A25D67F9AF   
25536  99F15E99-5A6E-4F9E-911B-C5F8E010BAD8   
25537  9B99C7C4-8160-46DC-89F7-C2F36B3767E2   
25538  9D1B7194-491E-42B4-A4AF-C1D4DBAA3CCE   
25539  B249150F-44BF-48CE-8135-D42FF59D0344   
25541  A397012C-D94E-45CA-A7E6-BD6F074B7028   
25544  AAAE0503-DE4B-47B6-8C4B-B46D0062D536   
25545  AAE3A59D-9701-4A7E-8EC4-BA03027E466E   
25547  ACBAEB96-C59A-433B-BC77-BAD807671AA4   
25548  AD07723C-4B8A-40DB-B09C-C83E40869CFD   
25549  AD1B5C1D-39E1-4A1F-A58F-D5F2DE64FE7F   
25550  AD4118A8-85A8-403C-8108-C3C6146D14C4   
25551  AD44F225-3942-418A-B036-C699233AD12A   
25552  ADCEEE0C-FB0E-4EED-89BB-C0FB6A253F0E   
25553  6E8646B8-5C71-4DE5-9164-FC5E68CAE56C   
25554  7345A886-BEC0-4B87-9EE2-E5B93F7938AF   
25555  6716AB88-0CC4-464D-9A37-FAEFDBF28FB7   
25557  6917202A-7D08-4660-964C-EF12D6E0F8E3   
25558  82F16418-FF71-4AB8-9185-FDBCE2CFB46E   
25559  861D241C-0324-4689-87EF-F91D681ECD16   
25560  86952E0A-D64A-4CE7-A0ED-EE2C1F02A07F   
25561  88C1DE3B-3852-4F78-8685-F80B4D748476   
25562  7AA4FAF1-DD16-466B-8852-EE37FB8A85F0   
25563  7CEF38BD-DE6D-45C1-9335-EC1B884E4146   
25564  7EC3733E-A3C2-4449-8889-F893E9C790E8   

                                              name        lat         lng  
0                      A-Tech Fabrications Limited  54.604095   -1.574308  
2                             Zurich Insurance plc  50.881574   -1.242902  
3                                 Limagrain UK Ltd  53.482100   -0.248484  
4                  China Earthquake Administration  39.977022  116.306198  
5                         Newmills Engineering Ltd  51.504765   -2.484122  
6                            Maastricht University  50.844288    5.688604  
8                                    Synthomer Ltd  51.784388    0.119136  
9                      Committee on Climate Change  51.494349   -0.140087  
10                          Agility Global Limited  51.650564   -0.149673  
11                                    Openreach BT  51.515741   -0.097966  
13                             Emote Games Limited  50.876386   -1.240271  
14                        Total Scientific Limited  52.094562    0.155632  
15              Omega Engineering Services Limited  51.901120   -2.097201  
18                               Swanshurst School  52.429668 

In [375]:
engine = create_engine('postgresql://{}:{}@{}/{}'.format(conf.get('user'),
                                                         conf.get('passw'),
                                                         conf.get('host'),
                                                         conf.get('database')))

conn = engine.connect()

In [483]:
df_result.to_sql('orgs_latlng',
                 conn,
                 schema='gtr',
                 if_exists='replace',
                 index=False)

# From here delete

In [32]:
latlng = pickle.load(open('latlng.p', 'rb'))

In [22]:
g = geocoder.geocode("SG8 7RE")
g

Location(Church St, Thriplow, Royston, Cambridgeshire SG8 7RE, UK, (52.0997269, 0.1046752, 0.0))

In [185]:
df_result = df.merge(df_latlng, left_index=True, right_index=True)
df_result.rename(columns={0: 'latlng'}, inplace=True)

In [281]:
df_result[~df_result.country.isin(uk_names)].country.unique()

array(['china', 'netherlands', 'australia', 'canada', 'austria', 'ireland',
       'japan', 'norway', 'brazil', 'sri lanka', 'switzerland', 'france',
       'united states', 'kenya', 'new zealand',
       'taiwan, province of china', 'zimbabwe', 'portugal', 'germany',
       'estonia', 'india', 'russian federation', 'singapore', 'nigeria',
       'spain', 'belgium', 'poland', 'italy', 'south africa', 'ethiopia',
       'bolivia, plurinational state of', 'denmark', 'uganda', 'viet nam',
       'kyrgyzstan', 'malawi', 'malaysia', 'mexico', 'argentina',
       'hungary', 'el salvador', 'bangladesh', 'greece', 'ukraine',
       'cambodia', 'chile', 'peru', 'thailand', 'mozambique', 'ghana',
       'israel', 'cuba', 'sierra leone', 'morocco', 'hong kong', 'zambia',
       'cameroon', 'tanzania, united republic of', 'sweden', 'iceland',
       'kazakhstan', 'korea, republic of', 'bosnia and herzegovina',
       'ecuador', 'finland', 'croatia', 'saudi arabia', 'madagascar',
       'slovenia',

In [254]:
df_result['country'] = df.addresses.map(country)
df_result['line1'] = df_result.addresses.map(line1)

In [255]:
df_other_countries = df_result[~df_result.country.map(lambda x: x.lower()).isin(uk_names)]

In [425]:
usa = ['usa', 'united states']
df_usa = df_other_countries[df_other_countries['country'].isin(usa)]

In [309]:
num_addresses = df_usa.iloc[166:].addresses.shape[0]
f = FloatProgress(min=0, max=num_addresses)
display(f)

for address in df_usa.iloc[166:].addresses:
    g = geocode(address)
    try:
        us_latlng.append((g.latitude, g.longitude))
    except AttributeError:
        us_latlng.append(None)
    f.value += 1

In [427]:
df_usa.reset_index(drop=True, inplace=True)
df_usa.drop('latlng', axis=1, inplace=True)
df_usa = df_usa.merge(df_latlng, left_index=True, right_index=True)
df_usa.rename(columns={0: 'latlng'}, inplace=True)

/usr/local/lib/python3.5/site-packages/ipykernel/__main__.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  from ipykernel import kernelapp as app


In [428]:
df_usa

id  \
0    2795AB81-FD33-413D-BBE4-127D1ACE6210   
1    55044D17-2C14-4D68-BED1-119271B476B4   
2    58E3A7F8-5B1E-48D2-BDC5-0E835B3FA948   
3    5FC4C032-4778-4A82-95CD-13F5897F599B   
4    527728F1-2397-4BCC-AA10-1137F55A1D9E   
5    1003BE2C-9E2D-4A5D-A13D-058587870722   
6    B3C4BA7F-0E8E-4F2F-88C2-04277E328EB0   
7    AFE5C6FD-3500-44F6-B100-184B5F2FD0D7   
8    C1DA4809-2264-4894-9C3C-190942EACF3C   
9    B8480401-8DDE-472E-9BFB-13DEC29BD816   
10   1B56BE39-25BE-4A95-9BAA-5F7EC425A638   
11   FB74D388-22C0-4DAB-9D15-BFF9CFA45E4B   
12   FC9ED2E5-3AE9-4CB4-AD0C-B8953F01D45C   
13   39313FCF-0990-4E36-B136-80AC7830128F   
14   F25BC0C5-5284-441F-B79F-A61707946CA0   
15   A1067C91-D7E3-4CBC-A671-583F17B73185   
16   99254C6D-7CCC-48F0-B118-7E25FF247891   
17   93589381-05D5-494F-875F-5DCE3BA590C0   
18   65E2BD5E-63AB-4245-BB34-8FF0EBA452A1   
19   6D30F26F-5158-46AA-AE3E-9D9EF1395FD8   
20   7266DA07-C495-49A1-B0A3-7C4C39A58D6A   
21   729C9072-B04E-45DD-A0B1-74028C36235E   
22   736985B4-EC0B-49E5-B8B0-7C91DBCFB55D   
23   5810877C-CDDA-4FFA-901C-6A9055CF8735   
24   4B4BAD27-8E4B-4D62-B7FB-6A3382E633AF   
25   E6756F63-3FA1-4DD9-8FFB-67C6EB0C4E90   
26   7852EBBB-C808-481F-8DAE-65F3E57E48CD   
27   88B1E177-B3F4-421A-85E1-683228DE3D70   
28   88D88D16-DC00-4266-AB15-99A0172E1310   
29   A486D494-6ED1-4FFC-A58E-804F8D5FDB7F   
..                                    ...   
569  483ABDB0-1754-410A-8FB0-A9A5CB2D0538   
570  520A7BCF-C90B-4FC1-A771-68E66D8A2810   
571  72C63E68-FE68-4BE1-B650-D43E6DB3B0AF   
572  C46CFCDF-00C3-43D7-B7A2-7C2ACB111142   
573  DEA494C1-5B42-42F7-9BF0-7BF5A402B897   
574  4EA14727-A869-4614-8185-82D83A9EEBF2   
575  56E4415F-A969-4B4E-9AE7-95CEB911461A   
576  74260C67-536C-43F4-802C-80C96A7F1C0C   
577  377F78A6-170E-4351-95BE-CEE25DB3519A   
578  A4FAAE42-7E74-4526-B17A-8AE5AE9B1EFB   
579  55B138DA-C696-4E77-B2CC-E7B5D95C06C1   
580  B1F0FBD2-C778-4BD5-B86A-88D96462C296   
581  D4B3BE36-563D-45F5-80FF-F7827271BB06   
582  78172E4D-9B35-4F46-B445-AB5B85DE5A37   
583  88D135CF-CE84-4C0E-8518-917AE24F1F71   
584  468C0716-C181-4A85-8827-B00269E6F90A   
585  02D51A91-D2E9-42DF-A97D-B562B52181FF   
586  31C4CAD2-D530-4D56-85C7-DD474F317B4F   
587  3849AA61-7794-4F3A-94FF-E246C05C6D8D   
588  436300CF-D086-49AD-8B44-CB277E1A1D95   
589  B3581EF6-1370-4D24-8B29-F17EC12D0E27   
590  B62A7F23-290D-4D84-B017-FBBDF6E49E3D   
591  CACF1F25-94D5-46CA-AF92-C83D9D0F43F3   
592  C103B777-2ADB-42DC-8302-D3891F569454   
593  8A05CB2F-89D2-4E3D-8357-E26DA1E4D1C8   
594  6E6F608F-23C5-4170-9198-B634BC9B7624   
595  57A137CC-2386-46FB-B01E-FBC25FC526D7   
596  43F70FE4-97E9-460D-BE06-F2FCD4B47699   
597  5C1B47CE-FF59-4725-8849-B3A4EDB98DF6   
598  84E231EE-EC82-44AC-9838-C30021DDA51A   

                                         name  \
0                       University of Chicago   
1    Rosenstiel School of Marine & Atmospheri   
2                           Xerox Corporation   
3                   Joint Bioenergy Institute   
4                                 Aginova Inc   
5                           Auburn University   
6                  Search for a Common Ground   
7                          Microsoft Research   
8                Mississippi State University   
9                  Louisiana State University   
10                                 Cardno TBE   
11                   East Carolina University   
12        University of Alabama at Birmingham   
13                         EcoHealth Alliance   
14       University of California Los Angeles   
15       University North Carolina Wilmington   
16                               Gaussian Inc   
17                                Synthes Inc   
18                        Genzyme Corporation   
19                                   Fermilab   
20                            Signal Patterns   
21                             Vassar College   
22               State University of New York   
23                          Mitre Corporation   
24               Dana F

In [435]:
df_usa.to_sql('orgs_latlng_us',
              conn,
              schema='gtr',
              if_exists='replace',
              index=False)